### ISSUE TITLE
Add Pair module to Stdlib

### ISSUE THREAD

**Issue by lupjo** [ocaml/ocaml#11651]
I think a module to handle pair would be welcome. I see these functions in it, maybe I'm forgetting a few more.
Also we could move `fst` and `snd` from stdlib.ml to pair.ml.

**pair.mli**
``` ocaml
type ('a, 'b) t = 'a * 'b

val equals: ('a -> 'a -> bool) -> ('b -> 'b -> bool) -> ('a, 'b) t -> ('a, 'b) t -> bool
val compare: ('a -> 'a -> int) -> ('b -> 'b -> int) -> ('a, 'b) t -> ('a, 'b) t -> int
val map: ('a -> 'c) -> ('b -> 'd) -> ('a, 'b) t -> ('c, 'd) t
val iter: ('a -> unit) -> ('b -> unit) -> ('a, 'b) t -> unit
val map2:
	('a -> 'c -> 'e) -> ('b -> 'd -> 'f) -> ('a, 'b) t -> ('c, 'd) t -> ('e, 'f) t
val swap: ('a, 'b) t -> ('b, 'a) t
val fold: ('a -> 'b -> 'c) -> ('a, 'b) t -> 'c
val create: 'a -> 'b -> ('a, 'b) t
```
**pair.ml**
``` ocaml
type ('a, 'b) t = 'a * 'b

let equals eqa eqb (a, b) (a', b')  =
	eqa a a' && eqb b b'
let compare cmpa cmpb (a, b) (a', b')  =
	 let c = cmp a a' in
	 if c <> 0 then c
	 else cmp b b'
let map f g (x, y) = (f x, g y)
let iter f g (x, y) = f x; g y
let map2 f g (x, y) (z, w) = (f x z, g y w)
let swap (x, y) = (y, x)
let fold f (x, y) = f x y
let create x y = (x, y)
```

EDIT: adding compare and iter

**Comment by xavierleroy** [comment]
It would be considerate to discuss this proposal.  I'm of two minds: on the one hand, OCaml has pairs but also triples, quadruples, etc, and it feels odd to favor one over the other, yet impossible to write generic functions over n-tuples; on the other hand, a few of the proposed functions look useful to me, such as `swap`, `equal` (no `s` please), and especially `compare`.

**Comment by nojb** [comment]
I agree that such a module would be useful. One could justify adding `Pair` on the basis that pairs are by far the most common among all tuples...

**Comment by lupjo** [comment]
Just found out there is actually a module [CCPair](https://ocaml.org/p/containers/3.13.1/doc/CCPair/index.html) in containers which contains the same functions I proposed and then some. I don't know if it's an argument in favour of not adding it to the stdlib or the opposite but I thought I'd update ;)

**Comment by yallop** [comment]
I think it'd be good to have this.  It's a nice complement to Either.

**Comment by nojb** [comment]
@lupjo would you want to give this a shot? There is no guarantee it will get merged, but I feel there is a soft consensus for this addition.

**Comment by lupjo** [comment]
Can do. I'll be in between contracts in a few weeks so I will get on that then

**Comment by Octachron** [comment]
The proposed module was implemented in #13310 and merged.
